util/progress/progressui/display.go

```diff
@@ -276,6 +285,134 @@ func (d *plainDisplay) done() {
 	d.t.printErrorLogs(d.t.w)
 }
 
+type jsonDisplay struct {
+	t    *trace
+	logs *logMux
+	enc  *json.Encoder
+}
+
+// newJSONDisplay creates a new SolveStatusDisplay that outputs the completed vertex
+// progress indicator for each vertex in JSON format.
+func newJSONDisplay(w io.Writer, opts ...DisplayOpt) Display {
+	enc := json.NewEncoder(w)
+	enc.SetIndent("", "  ")
+	return Display{
+		disp: &jsonDisplay{
+			t:    newTrace(w, false),
+			logs: newLogMux(),
+			enc:  enc,
+		},
+	}
+}
+
+func (d *jsonDisplay) init(displayLimiter *rate.Limiter) {
+	// Display limiter is ignored for the JSON display.
+}
+
+func (d *jsonDisplay) update(ss *client.SolveStatus) {
+	d.t.update(ss, 80)
+
+	// JSON format keeps the logs separated into their streams
+	// and takes the aggregated result. The trace does not do that
+	// so we use and update logMux for log messages by digest/stream.
+	for _, l := range ss.Logs {
+		d.logs.update(l)
+	}
+
+	// Only look at the digests that have updates.
+	// If a digest doesn't have an update, it can't
+	// have been completed.
+	for dgst := range d.t.updates {
+		v, ok := d.t.byDigest[dgst]
+		if !ok {
+			continue
+		}
+
+		if v.isCompleted() {
```

Comment:
Looking at this again, would the case that all blobs are shown separately be solved if in here we instead check that all the `mergedIntervals` are completed, then in `printVtx` we combine the (unprinted) `mergedIntervals` and reset the list after printing to discard them.

Reply:
From what I can see, `mergedIntervals` only contains the one interval during this time. From what I see, we only know that the current vertex has a status that is completed and we don't see any in progress statuses at this point.